softioc/device.py

```diff
@@ -323,35 +375,43 @@ def _write_value(self, record, value):
             self.dtype.itemsize * nord)
         record.NORD = nord
 
+    def _handle_strings(self, value):
+        """Handles strings and bytearrays as values for Waveforms. Returns the
+        string converted to a numpy array."""
 
-class waveform(WaveformBase, ProcessDeviceSupportIn):
-    _record_type_ = 'waveform'
-    _device_name_ = 'devPython_waveform'
-    _default_ = ()
-
-    # Because arrays are mutable values it's ever so easy to accidentially call
-    # set() with a value which subsequently changes.  To avoid this common class
-    # of bug, at the cost of duplicated code and data, here we ensure a copy is
-    # taken of the value.
-    def set(self, value,
-            severity=alarm.NO_ALARM, alarm=alarm.UDF_ALARM, timestamp=None):
-        '''Updates the stored value and triggers an update.  The alarm
-        severity and timestamp can also be specified if appropriate.'''
         if isinstance(value, str):
+            value = value.encode(errors="replace")
+
+        if isinstance(value, bytes):
             # Convert a string into an array of characters.  This will produce
             # the correct behaviour when treating a character array as a string.
             # Note that the trailing null is needed to work around problems with
-            # some clients.
-            value = numpy.fromstring(value + '\0', dtype = 'uint8')
+            # some clients. Note this also exists in builder.py's _waveform().
+            value = numpy.frombuffer(
+                value + b'\0',
+                dtype = numpy.uint8
+            )
 
-        value = numpy.require(value, dtype = self.dtype)
-        self._value = (+value, severity, alarm, timestamp)
-        if value.shape == ():
-            value.shape = (1,)
-        assert value.ndim == 1, 'Can\'t write multidimensional arrays'
+        return value
```

Comment:
So what happens to values that don't pass either `isinstance` test?  Just passed through?

That's what the `numpy.require` was about, and it does the appropriate `.dtype` conversion here.

Or is that all subsumed into the `value_to_dbr` call that follows on?

Reply:
Anything that isn't a `bytes` or a `str` will just pass through. This method is not supposed to touch them. I can move those checks out of the method if you prefer, although that does end up duplicating the `bytes` check. 

The diff makes this look pretty weird. The `numpy.require` does still happen, it's inside the `value_to_dbr` call. This method is just some pre-processing before that method.